package/src/cpumon.cpp

```diff
@@ -63,60 +63,60 @@ std::map<std::string, double> const cpumon::get_json_average_stats(
+  // Map lscpu names to the desired ones in the JSON
+  const std::map<std::string, std::string> metricToName{
+      {"Model name", "ModelName"},
+      {"CPU(s)", "CPUs"},
+      {"Socket(s)", "Sockets"},
+      {"Core(s) per socket", "CoresPerSocket"},
+      {"Thread(s) per core", "ThreadsPerCore"}};
+
+  // Useful function to determine if a string is purely a number
+  auto isNumber = [](const std::string& s) {
+    return !s.empty() && std::find_if(s.begin(), s.end(), [](unsigned char c) {
+                           return !std::isdigit(c);
+                         }) == s.end();
+  };
+
+  // Loop over the output, parse the line, check the key and store the value if
+  // requested
+  std::string key{}, value{};
+
+  for (const auto& line : cmd_result.second) {
+    // Continue on empty line
     if (line.empty()) continue;
+
+    // Tokenize by ":"
     size_t splitIdx = line.find(":");
-    std::string val;
-    if (splitIdx != std::string::npos) {
-      val = line.substr(splitIdx + 1);
-      if (val.empty()) continue;
-      for (const auto& metric : metrics) {
-        if (line.size() >= metric.size() &&
-            line.compare(0, metric.size(), metric) == 0) {
-          if (metric == "processor")
-            nCPU++;
-          else if (metric == "siblings") {
-            if (nSiblings == 0) nSiblings = std::stoi(val);
-          } else if (metric == "cpu cores") {
-            if (nCores == 0) nCores = std::stoi(val);
-          } else if (hw_json["HW"]["cpu"][metric].empty()) {
-            hw_json["HW"]["cpu"][metric] =
-                std::regex_replace(val, std::regex("^\\s+|\\s+$"), "");
-          }
-        }  // end of metric check
-      }    // end of populating metrics
-    }      // end of seperator check
-  }        // end of reading cpuInfoFile
-
-  // The clear assumption here is that there is a single type of processor
-  // installed nCPU = nSockets * nSiblings nSiblings = nCoresPerSocket *
-  // nThreadsPerCore
-  // Next 4 lines are a workaround for Raspberry Pi
-  if (!nSiblings) nSiblings = nCPU;
-  if (!nCores) nCores = nCPU;
-  unsigned int nSockets = nCPU / nSiblings;
-  unsigned int nThreads = nSiblings / nCores;
-  hw_json["HW"]["cpu"]["nCPU"] = nCPU;
-  hw_json["HW"]["cpu"]["nSockets"] = nSockets;
-  hw_json["HW"]["cpu"]["nCoresPerSocket"] = nCores;
-  hw_json["HW"]["cpu"]["nThreadsPerCore"] = nThreads;
-
-  // Close the file
-  cpuInfoFile.close();
+    if (splitIdx == std::string::npos) continue;
+
+    // Read "key":"value" pairs
+    key = line.substr(0, splitIdx);
+    value = line.substr(splitIdx + 1);
+    if (key.empty() || value.empty()) continue;
+    key = std::regex_replace(key, std::regex("^\\s+|\\s+$"), "");
+    value = std::regex_replace(value, std::regex("^\\s+|\\s+$"), "");
+
+    // Fill the JSON with the information
+    if (metricToName.count(key) == 1) {
+      if (isNumber(value))
+        hw_json["HW"]["cpu"][metricToName.at(key)] = std::stoi(value);
```

Comment:
Isn't `metricToName[key]` better? Although `at` gives a bounds check, we know already that the key is present. And if it weren't then it would throw `std::out_of_range` that we don't catch and the program would crash anyway, so it's no protection.

Reply:
operator `[]` is not declared as const for `std::map` (or `std::unordered_map` for that matter), so that gives me a compilation error since the method is `cont`. That's why I ended up using `at`:

```
/home/amete/github/prmon/package/src/cpumon.cpp: In member function 'virtual const void cpumon::get_hardware_info(nlohmann::json&)':
/home/amete/github/prmon/package/src/cpumon.cpp:117:46: error: passing 'const std::unordered_map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >' as 'this' argument discards qualifiers [-fpermissive]
  117 |         hw_json["HW"]["cpu"][metricToName[key]] = value;
```

Let me know if I'm missing something here.

Reply:
You are right - I missed that! `[]` is also used a setter so it's not `const`. Yes, just leave it as is. Sorry for the noise!